### PR TITLE
[wip] Second test of render hooks for links

### PR DESCRIPTION
**Pull request by kbhawkey**
Testing Hugo Render Hooks to update links.
Testing changes to feature state (raw html not rendered).
For background information, see prs #19526, #19830.

### CODE REVIEW

**Comment by kbhawkey** [comment]
Note: There is debugging info inserted before links.

File: layouts/_default/_markup/render-link.html [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+{{- else if (fileExists $filePathHasLangPrefixNoSuffix) -}}
+<a href="{{ $link | relLangURL | safeURL }}"{{ with .Title}} title="{{ . }}"{{ end }}>USE TRANSLATED:file-exists {{- $filePathHasLangPrefixNoSuffix -}}{{ .Text | safeHTML }}</a>
+{{- else -}}
+<a href="{{ $link | safeURL }}"{{ with .Title}} title="{{ . }}"{{ end }}>DEFAULT to EN, file does not exist for {{- $link -}}{{ .Text | safeHTML }}</a>
+{{- end -}}
+{{- else -}}
+<a href="{{ $link | safeURL }}"{{ with .Title}} title="{{ . }}"{{ end }}>EN: {{- $filePath -}} {{ .Text | safeHTML }}</a>
+{{- end -}}
+{{- end -}}
+{{- end -}}
```
Comment (tengqm) [inline layouts/_default/_markup/render-link.html:37]: Tried to digest the logic from line 20 to 37:

```
{{ if $onlyFragment }}
    {{/* skip labels in current page */}}
    <a href="{{ $link | safeURL }}"{{ with .Title}} title="{{ . }}"{{ end }}>{{ .Text | safeHTML }}</a>
{{ else if $isRemote }}
    {{/* open remote page in new window/tab */}}
    <a href="{{ $link | safeURL }}"{{ with .Title}} title="{{ . }}" target="_blank"{{ end }}>{{ .Text | safeHTML }}</a>
{{ else if eq "en" $pageLang }}
    {{/* skip other English pages */}}
    <a href="{{ $link | safeURL }}"{{ with .Title}} title="{{ . }}"{{ end }}>{{ .Text | safeHTML }}</a>
{{ else if or (fileExists $filePath) (fileExists $filePathHasLangPrefix) (fileExists $filePathNoSuffix) (fileExists $filePathHasLangPrefixNoSuffix) }}
    {{/* There is a translated version */}}
    {{/* THIS CHECK IS NOT SUFFICIENT FOR REDIRECTED LINKS */}}
    {{/* NOT SURE IF THIS WORKS FOR LINKS TO API REFERENCE URL WHICH CONTAINS GO TEMPLATES ALREADY */}}
    <a href="{{ $link | relLangURL | safeURL }}"{{ with .Title}} title="{{ . }}"{{ end }}>{{ .Text | safeHTML }}</a>
{{ else }}
    {{/* fall back to English version */}}
    {{/* IS FALLING BACK TO ENGLISH ALWAYS THE BEST OPTION? */}}
    <a href="{{ $link | safeURL }}"{{ with .Title}} title="{{ . }}"{{ end }}>{{ .Text | safeHTML }}</a>
{{ end }}
```

File: layouts/_default/_markup/render-link.html [hunk 2]
```diff
@@ -0,0 +1,37 @@
+{{- $link := .Destination }}
+{{- $isRemote := strings.HasPrefix $link "http" }}
+{{- $onlyFragment := strings.HasPrefix $link "#" }}
+{{- $pageLang := .Page.Lang -}}
+
+{{- $filePath := printf "content/%s/%s.md" $pageLang (trim $link "/" ) -}}
+{{- $filePathHasLangPrefix := printf "content/%s.md" (trim $link "/") -}}
+{{- $filePathNoSuffix := printf "content/%s/%s" $pageLang (trim $link "/") -}}
+{{- $filePathHasLangPrefixNoSuffix := printf "content/%s" (trim $link "/") -}}
```
[the remark this answers is not in the document]
Comment (kbhawkey) [reply layouts/_default/_markup/render-link.html:9]: @tengqm , Thanks for reviewing this work-in-progress PR.
Yes, handling the various cases is a challenge.
I am trying to understand the problems with updating links in localized pages (authors update page links based upon whether a linked page is localized; authors clean up stale links; updating links is hard).
What steps could be taken to make the link changes easier?
For now, the site has a mix of localized and English pages.
For each localization, the left hand navigation lists a combination of pages that are localized and not localized.

Q: Given these conditions, should a custom template (Hugo) publish updated links for localized pages (based upon the localization)? I guess not.

Perhaps what is needed is a way to _check_ the links on a localized page (such as a custom template or script run in debug mode), and have Hugo fail/warn in the build. The checks would point out that a localized page exists for the link and that the link needs to be updated to this localized page (do not use the `en` link, use the localized link). The script could list the links that need to be updated. The author corrects the links.
What do you think?

File: layouts/_default/_markup/render-link.html (hunk 1 above)
Comment (kbhawkey) [reply layouts/_default/_markup/render-link.html:37]: Yes, testing/debugging here.

**Pull request closed without merging**